tokio-stream/src/stream_ext.rs

```diff
@@ -993,6 +995,53 @@ pub trait StreamExt: Stream {
     {
         throttle(duration, self)
     }
+
+    /// Collects items into batches inside a vector within a deadline.
+    ///
+    /// Attempts to yield a vector of len `capacity` within the deadline,
+    /// otherwise will yield a vector of len less than `capacity` if the deadline is reached.
+    /// If the wrapped stream does not yield a value before the deadline is reached, then an
+    /// empty vector will be yielded. The deadline will be reset after every batch is yielded.
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// use std::time::Duration;
+    /// use tokio::time;
+    /// use tokio_stream::{self as stream, StreamExt};
+    /// use futures::FutureExt;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let iter = vec![1, 2, 3].into_iter();
+    ///     let stream0 = stream::iter(iter);
+    ///
+    ///     let iter = vec![4].into_iter();
+    ///     let stream1 = stream::iter(iter)
+    ///          .then(move |n| time::sleep(Duration::from_secs(5)).map(move |_| n));
+    ///
+    ///     let chunk_stream = stream0
+    ///         .chain(stream1)
+    ///         .chunks_timeout(3, Duration::from_secs(2));
+    ///
+    ///     // a full batch was received
+    ///     assert_eq!(chunk_stream.next().await, Some(vec![1,2,3]));
+    ///     // deadline was reached before capacity was reached
+    ///     assert_eq!(chunk_stream.next().await, Some(vec![4]));
+    ///     // no items were emitted within the deadline
+    ///     assert_eq!(chunk_stream.next().await, Some(vec![]));
```

Comment:
I was wondering something: maybe it would be more useful to change the stream so that the timer only starts counting down when the first item of the batch arrives?

Reply:
yeah that's a great point, i will update accordingly